test/release/examples/patterns/readcsv.chpl

```diff
@@ -0,0 +1,286 @@
+  var valDomain = {0..dataRows.size-1}; // Domain indicating how many values per column.
+  var colNameDomain: domain(string);    // An associative domain of strings.
+  for colName in colNames do            // Put each column name into the column name domain.
+    colNameDomain += colName;
+  // colData is an associative array indexed by the column name domain with array values.
+  // Each array value will have number of rows string elements.
+  var colData: [colNameDomain] [valDomain] string;
+
+  // Processing each row to put the column values into the associative array.
+  for rowIdx in 0..dataRows.size-1 {
+    // Pull the line out from the list of dataRows.
+    line = dataRows[rowIdx];
+    var start = 0;      // index into the line string
+    var colIdx = 0;     // the first column is indexed as zero
+    // Process each line by finding the next field value from the start index to the next comma.
+    while (start<line.size && colIdx<colNames.size) {
+      var (nextVal,nextCommaIdx) = nextField(line,start);
+      colData[colNames[colIdx]][rowIdx] = nextVal;
+      start = nextCommaIdx + 1;         // The next column of data will start after the comma.
+      colIdx = colIdx + 1;              // Increment the column index.
+    }
+  }
+  reader.close();
+
+  // Print out all of the data values for the first column of data.
+  if debug then writeln("\ncolData[colNames[0]]=",colData[colNames[0]]);
+
+  //-----------------------------------------------------------------------
+  // To see how the data is accessed and for testing, let's generate a 
+  // temporary csv file from the stored data.  It should be the same as the 
+  // input csv file.
+  if debug {
+    var outfile = open(approach2FileName, iomode.cw);
+    var writer = outfile.writer();
+
+    // First write to the output file the column names separated by commas.
+    for colIdx in 0..colNames.size-2 {
+      writer.write(colNames[colIdx],",");
+    }
+    // The last column name should not be followed by comma.
+    writer.writeln(colNames[colNames.size-1]);  
+
+    // Then write to the file each row of data.
+    for rowIdx in 0..colData[colNames[0]].size-1 {
+      for colIdx in 0..colNames.size-2 {
+        writer.write(colData[colNames[colIdx]][rowIdx],",");
+      }
+      // The last column value should not be followed by comma.
+      writer.writeln(colData[colNames[colNames.size-1]][rowIdx]);
+    }
+    writer.close();
+
+    // Compare the file generated from the associative array with the original input csv file.
+    checkThatFilesMatch(inFileName,approach2FileName);
+  }
+}
+
+// Returns the index of the next comma or the string length if no
+// next comma is found.  Commas inside quoted strings will be ignored.
+proc findNextCommaNotInQuotes(str : string, start : int) {
+  var inQuotes = false;
+  for i in start..(str.size-1) {
+    if !inQuotes && str[i]==',' then return i;
+    if str[i]=='"' then inQuotes = !inQuotes;
```

Comment:
Probably not important, but we care only about double quotes in the file and not single quotes.

Reply:
Clarified by renaming `inQuotes` to `inDoubleQuotes`.